### PR TITLE
add avoid_keyword_to_create_instances

### PR DESCRIPTION
**Pull request by a14n**
Fixes dart-lang/sdk#32602

### CODE REVIEW

**Comment by a14n** [comment]
Not mergeable until analyzer is upgrade.

**Comment by bwilkerson** [comment]
The tests will need to be updated to reflect the new semantics, but the implementation should not need to change.

**Comment by a14n** [comment]
PTAL

File: test/_data/avoid_keyword_to_create_instances/a.dart [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  A(); // OK
+
+  new A.c1(); // LINT
+  A.c1(); // OK
+
+  const A([]); // OK
+  new A([]); // LINT
+  A([]); // OK
+  A(const []); // OK
+  const A([]); // OK
```
Comment (bwilkerson) [inline test/_data/avoid_keyword_to_create_instances/a.dart]: This is a duplicate of line 20.

Comment (a14n) [reply test/_data/avoid_keyword_to_create_instances/a.dart]: removed. Thanks!

**Comment by bwilkerson** [comment]
Based on discussions elsewhere (let me know if you haven't seen them and I'll try to get you a pointer to them) I'm wondering whether we might want to split this into two lints: one to advise removing `const` inside a constant context and one to advise removing `new` where the constructor being invoked is not `const`.

**Comment by a14n** [comment]
>I'm wondering whether we might want to split this into two lints

I can do that but I don't think people will use only one of them.

**Comment by bwilkerson** [comment]
You might be right. We just don't have enough experience using the feature to know, at this point, what people are going to want.

**Comment by eernstg** [comment]
I think @bwilkerson was referring to [this comment](https://github.com/dart-lang/sdk/issues/32798#issuecomment-380043155), about reasons for allowing `new` in some situations.

**Comment by bwilkerson** [comment]
I was, yes. (I just didn't have a convenient way to find it at the time I wrote the comment.)

**Comment by a14n** [comment]
This lint was starting with tests that conformed to _magic const_ . As you can see in https://github.com/dart-lang/linter/pull/936/commits/3962014698ad3cb76634bebaf2683451567dedad#diff-d85039c777133b1ebb3ecb9c9366210fR14 (first version of this lint with magice const) `new A()` was not linted because it is not the same with or without `new`. Actually only the analyzer has changed to remove support of magic const and if/once magic const comes back the lint should magically handle that ( it's all about magic after all :D )

**Comment by a14n** [comment]
PTAL

**Comment by a14n** [comment]
Reminder: I can't merge :)

**Comment by bwilkerson** [comment]
I actually did remember, this time. I was waiting to see whether @eernstg has any additional thoughts.

**Comment by bwilkerson** [comment]
I'm guessing he didn't. Should I merge this now, or are there other changes you wanted to make first?

**Comment by a14n** [comment]
I think it's ok. Trying this lint on flutter I faced flutter/flutter#17429 but it looks like a bug in analysis server. Wdyt?

**Comment by bwilkerson** [comment]
@devoncarew @pq Should we wait to land this until we've found and fixed the referenced issue so that we'll know whether they're the same or different issues?

**Comment by pq** [comment]
I don't feel strongly, @devoncarew ?

**Comment by devoncarew** [comment]
We should figure out whether this will break the next flutter roll; I asked a follow up there but have not investigated the failure.

**Comment by devoncarew** [comment]
We can merge no? We just can't enable this lint on the flutter repo until the other issues are address.

**Comment by bwilkerson** [comment]
Yes, we can merge. But there's no way for users to know that the lint shouldn't be used until other issues have been fixed.

**Comment by pq** [comment]
Sorry to have let this slip.  @a14n: are we still wanting to land this as is?  Any updates?

**Comment by a14n** [comment]
Removing `new` causes issues with `prefer_const_literals_to_create_immutables` and `prefer_const_constructors` so it could be worth to split this lint into 2 lints (one for remonving `new` and an other for removing `const`). Thus we could apply the removing of `const` without waiting the fix on ConstantVerifier. I don't know when @bwilkerson plans to look at a fix for ConstantVerifier.

**Comment by pq** [comment]
OK, cool.  Thanks @a14n.  That's what I thought.  @bwilkerson is on vacation; let's regroup when he's back. 👍

**Comment by bwilkerson** [comment]
I don't know when I will be able to look at that issue. Of course, anyone else is also welcome to look at it. :-)

**Comment by pq** [comment]
@a14n: this looks great.  Thanks for the follow-through!

**Pull request merged**